mlflow/pytorch/__init__.py

```diff
@@ -736,56 +742,107 @@ def get_raw_model(self):
         """
+        Overrides the predict method to handle dict and list of dict inputs.
+
         Args:
-            data: Model input data.
-            params: Additional parameters to pass to the model for inference.
+            data: Input data, which can be pandas DataFrame, numpy ndarray,
+                  dict of numpy arrays, or list of dicts of numpy arrays.
+            params: Additional parameters for inference (currently not used).
 
         Returns:
-            Model predictions.
+            Model predictions as numpy arrays or pandas DataFrames.
         """
-        import torch
 
         if params and "device" in params:
             raise ValueError(
-                "device' can no longer be specified as an inference parameter. "
+                "'device' can no longer be specified as an inference parameter. "
                 "It must be specified at load time. "
                 "Please specify the device at load time, for example: "
                 "`mlflow.pyfunc.load_model(model_uri, model_config={'device': 'cuda'})`."
             )
 
-        if isinstance(data, pd.DataFrame):
-            inp_data = data.values.astype(np.float32)
-        elif isinstance(data, np.ndarray):
-            inp_data = data
-        elif isinstance(data, (list, dict)):
-            raise TypeError(
-                "The PyTorch flavor does not support List or Dict input types. "
-                "Please use a pandas.DataFrame or a numpy.ndarray"
-            )
-        else:
-            raise TypeError("Input data should be pandas.DataFrame or numpy.ndarray")
+        input_tensor = self._preprocess_input(data)
 
-        device = self.device
-        with torch.no_grad():
-            input_tensor = torch.from_numpy(inp_data).to(device)
+        # Perform inference
+        with self.torch.no_grad():
             preds = self.pytorch_model(input_tensor)
-            # if the predictions happened on a remote device, copy them back to
-            # the host CPU for processing
-            if device != _TORCH_CPU_DEVICE_NAME:
-                preds = preds.to(_TORCH_CPU_DEVICE_NAME)
-            if not isinstance(preds, torch.Tensor):
-                raise TypeError(
-                    "Expected PyTorch model to output a single output tensor, "
-                    f"but got output of type '{type(preds)}'"
-                )
+
+        return self._postprocess_output(preds, data)
+
+    def _preprocess_input(self, data: Any):
+        """
+        Preprocesses input data to convert it into torch.Tensors suitable for the model.
+
+        Args:
+            data: Input data.
+
+        Returns:
+            Torch tensor or dictionary of torch tensors suitable for model input.
+        """
+        device = self.device
+
+        if isinstance(data, np.ndarray):
+            # Convert ndarray to tensor
+            return self.torch.from_numpy(data.astype(np.float32)).to(device)
```

Comment:
I see... My mistake. I will correct all this.. Regarding supporting flexible dtype,  I suppose it is not such a big problem since Json does  support float32. But it may be interesting. I was hoping to add binary inputs at some point, the binary type comes up quite handy for complex data types such as parquet files. Maybe I am ignoring something but the way around I found was encoding all as base64.  But in general,  I end up loosing resolution, bigger payload sizes and more encoding/decoding time. What do you think?